flytekit/core/promise.py

```diff
@@ -36,6 +37,33 @@
 from flytekit.models.types import SimpleType
 
 
+def _is_none_valid_instance_of(t: Type) -> bool:
+    """
+    Returns whether None is a valid instance of t.
+
+    >>> _is_none_valid_instance_of(type(None))
+    True
+
+    >>> _is_none_valid_instance_of(type(int))
+    False
+
+    >>> _is_none_valid_instance_of(Union[int, str])
+    False
+
+    >>> _is_none_valid_instance_of(Optional[int])
+    True
+
+    >>> _is_none_valid_instance_of(Optional[List[int]])
+    True
+
+    >>> _is_none_valid_instance_of(Union[str, Optional[int]])
+    True
+    """
+    if get_origin(t) is None:
+        return t is type(None)
+    return any(_is_none_valid_instance_of(arg) for arg in get_args(t))
```

Comment:
`get_args` is [generic](https://docs.python.org/3/library/typing.html#typing.get_args) right?  It's used not just for Union, is there a way to scope this call to just look at Union types?  a bit worried that this check might cause a false positive somehow.